### ISSUE TITLE
[Bug] websocket error with CRA 4.0.1 and storybook 6

### ISSUE THREAD

**Issue by MaksimShakavin** [storybookjs/presets#177]
### Describe the bug

After updating to storybook 6 and CRA 4 I see websocket errors in browser console and dev-server constantly disconnects.

`WebSocket connection to 'ws://localhost:9000/sockjs-node' failed: Error during WebSocket handshake: Unexpected response code: 404`

### Steps to reproduce the behavior

1. Create empty CRA project with the latest version (4.0.1)
2. Init it with storybook 
3. See error in opened browser

### Expected behavior

Hot reloading is working and there are no errors. 

### Screenshots and/or logs

`Request URL: ws://localhost:9000/sockjs-node`

### Environment

 - OS: MacOs 
 - Node.js version: 12.12.0
 - NPM version: 6.11.13
 - Browser (if applicable): chrome

**Comment by Silon** [comment]
@MaksimShakavin any updates? Hve you solved this issue?

**Comment by edoardo-bluframe** [comment]
Hey guys!

Getting the **very** same error here!

* CRA: 4.0.1
* Storybook: 6.1.14

Do we have an idea as to how we fix it?

**Comment by stephenh** [comment]
We're also seeing this. Before upgrading it looks like there was a single `react-dev-utils` package:

```
$ npm ls react-dev-utils
internal-frontend@0.1.0 /home/stephen/homebound/internal-frontend2
├─┬ @storybook/addon-docs@6.0.21
│ └─┬ @storybook/core@6.0.21
│   └── react-dev-utils@10.2.1  deduped
├─┬ UNMET PEER DEPENDENCY @storybook/react@6.0.21
│ └── react-dev-utils@10.2.1
└─┬ react-scripts@3.4.1
  └── react-dev-utils@10.2.1  deduped
```

But after upgrading to cra-4, it looks like cra-4 brought in a new react-dev-utils 11.0.0:

```
$ npm ls react-dev-utils
internal-frontend@0.1.0 /home/stephen/homebound/internal-frontend
├─┬ @storybook/addon-docs@6.0.21
│ └─┬ @storybook/core@6.0.21
│   └── react-dev-utils@10.2.1  deduped
├─┬ UNMET PEER DEPENDENCY @storybook/react@6.0.21
│ └── react-dev-utils@10.2.1
└─┬ react-scripts@4.0.0
  └── react-dev-utils@11.0.0
```

Large disclaimer that `react-dev-utils` may have nothing to do this with, I just noticed the difference while trying to divine "what changed between working & not-working".

**Comment by stephenh** [comment]
Well, looks like my guess about `react-dev-utils` was wrong; it got bumped to the latest by @thebuilder in:

https://github.com/storybookjs/storybook/commit/efdb20b68a5e44c0e21c3cdb14c98ae0fd7a0d75

But trying `6.2.0-alpha.14` locally (and verifying I only have a single version of `react-dev-utils`), I still get the same issue...

I'm pretty stuck on knowing how to debug this. 

@thebuilder or @ghengeveld do you have any suggestions for how we could debug this? Or could you @ in someone who's an expert on the HMR functionality of storybook?

Thanks!

**Comment by stephenh** [comment]
@shilman & @ndelangen apologies for the at but looks like you've fixed HMR issues before...do you have any ideas/hints for this?

I've created a repository with a repro:

https://github.com/stephenh/cra-playground

All I did was 1) install latest cra, 2) install storybook, both with the out-of-the-box `npx` commands, and I immediately get the:

"WebSocket connection to 'ws://192.168.86.187:6006/sockjs-node' failed: Error during WebSocket handshake: Unexpected response code: 404"

issue.

**Comment by evcheung** [comment]
Bumping this, I'm currently experiencing the same issue on Storybook 6.1.14

**Comment by stephenh** [comment]
Fwiw I tried running the latest Storybook 6.2.0-alpha.18 and still broken.

**Comment by RanzQ** [comment]
Setting FAST_REFRESH=false seems to help: https://create-react-app.dev/docs/advanced-configuration/

**Comment by Vagabun** [comment]
same issue, currently experiencing it with storybook-react@6.1.11

**Comment by designorant** [comment]
Can confirm with `create-react-app@4.0.2` and `@storybook/react@6.1.16`. Setting `FAST_REFRESH=false` does not help.

**Comment by DouglasDev** [comment]
I can confirm this is happening with react-scripts 4.0.1 and storybook/react 6.1.14. Strangely HMR works fine if I change the code in a .stories.js file, but not if I change any of the files that are imported into the story. If someone can point me in the right direction I would be willing to attempt a fix. Anyone know which file in the storybook repo deals with reloading/HMR?

**Comment by travis5491811** [comment]
I'm having this issue with `@storybook/vue': '^6.1.14'`. I'm also willing to try and fix. I just don't know where to start with troubleshooting a issue with a dependency of a dependency (project, storybook, [webpack])

**Comment by shilman** [comment]
@stephenh I am seeing the error `ws://localhost:9000/sockjs-node` and will continue to investigate.

That said, HMR appears to be 100% functional in the apps I've tested, for both `latest` and `next` versions of Storybook, as well as for your repro repo.

Is HMR _actually_ broken for you, or are you just bothered by this `sockjs-node` error?

**Comment by stephenh** [comment]
@shilman it _was_ actually broken yes, however when checking this morning, there is now very obvious console output showing that `xhr-mock` is intercepting the HMR requests and aborting them.

We have a few stories that use a decorator to install `xhr-mock` and if I just comment that `mock.setup()` call out, everything magically works again (other than those stories themselves, we'll have to work on selectively getting xhr-mock back or using a different technique).

So, a few curiosities:

1) Before CRA 4.x, somehow we were able to use `xhr-mock` for stories and it was not intercepting the HMR requests.

2) I swear on our first CRA 4.x upgrade (to cra `4.0.1` with storybook `6.1.14`), the only console output we saw was `/sockjs-node`. But then ~a week ago another dev on the team bumped us to even-newer-CRA-4 (from `4.0.1` -> `4.0.3`) and even-newer-storybook (from `6.1.14` to `6.1.21`), specifically to fix this issue, and while hot reloading was still broken, I assume that is when this the very-helpful/obvious `xhr-mock` console errors started showing up to help ~almost instantly diagnose the issue.

I guess there are a few alternatives here:

a) is that on cra-4.0.1 and sb-6.1.14 this `xhr-mock` was the issue and I was just blind and didn't see it
b) there was some other bug/mismatch in cra-4.0.1 and sb-6.1.14 that has since been magically fixed, and so now the `xhr-mock` error shows up and is easily fixed

But, regardless, yes, even though I see a `/sockjs-node` error, I have hot reload working! Thank you @shilman for the nudge to just check again!

Would be curious if others on this thread are also using `xhr-mock` for stories and if anyone has ideas about what changed in cra-4.x such that `xhr-mock` became an issue...

**Comment by stephenh** [comment]
Thinking about it, I bet the `/sockjs-node` error is HMR trying to work via websockets, and on cra-3.x that worked fine, so that is why `xhr-mock` didn't matter, b/c it only intercepts AJAX, and the HMR was going through websockets.

Then, whatever broke websockets in the combination of cra-4.x + storybook-6.1, (probably?) means that HMR is falling back on AJAX polling, and that is why `xhr-mock` getting in the way was now an issue.

**Comment by shilman** [comment]
Fixed and released in 3.1.7
